File: apps/mobile/ios/Happy/Services/PurchaseService.swift
//
//  PurchaseService.swift
//  Happy
//
//  Copyright (c) 2024-2026 Enflame Media. All rights reserved.
//

import Foundation
import StoreKit

// MARK: - PurchaseProviding Protocol

/// Protocol defining the purchase service interface.
///
/// This abstraction allows swapping between StoreKit 2 (native) and
/// RevenueCat implementations. The default implementation uses StoreKit 2
/// directly. When RevenueCat SDK is integrated, create a conforming
/// `RevenueCatPurchaseService` class that wraps the RC SDK calls.
protocol PurchaseProviding: AnyObject {
    /// The current subscription status.
    var subscriptionStatus: SubscriptionStatus { get }

    /// Available subscription plans loaded from StoreKit.
    var availablePlans: [SubscriptionPlan] { get }

    /// Load available products from the App Store.
    func loadProducts() async throws

    /// Purchase a subscription plan.
    /// - Parameter plan: The plan to purchase.
    /// - Returns: Whether the purchase was successful.
    @discardableResult
    func purchase(_ plan: SubscriptionPlan) async throws -> Bool

    /// Restore previously purchased subscriptions.
    func restorePurchases() async throws

    /// Check and update the current subscription status.
    func refreshSubscriptionStatus() async

    /// Whether a purchase is currently in progress.
    var isPurchasing: Bool { get }
}

// MARK: - Purchase Errors

/// Errors that can occur during purchase operations.
enum PurchaseError: LocalizedError, Equatable {
    case productNotFound(String)
    case purchaseFailed(String)
    case purchaseCancelled
    case verificationFailed
    case networkError(String)
    case storeKitError(String)
    case notAvailable
    case pendingTransaction

    var errorDescription: String? {
        switch self {
        case .productNotFound(let id):
            return "Product '\(id)' not found in the App Store."
        case .purchaseFailed(let reason):
            return "Purchase failed: \(reason)"
        case .purchaseCancelled:
            return "Purchase was cancelled."
        case .verificationFailed:
            return "Transaction verification failed. Please try again."
        case .networkError(let reason):
            return "Network error during purchase: \(reason)"
        case .storeKitError(let reason):
            return "App Store error: \(reason)"
        case .notAvailable:
            return "In-app purchases are not available on this device."
        case .pendingTransaction:
            return "A transaction is pending approval (e.g., Ask to Buy)."
        }
    }
}

// MARK: - PurchaseService (StoreKit 2 Implementation)

/// Native StoreKit 2 purchase service for managing subscriptions.
///
/// Uses async/await APIs from StoreKit 2 to handle product loading,
/// purchasing, and subscription status tracking. Conforms to `PurchaseProviding`
/// for easy swapping with a RevenueCat-backed implementation later.
///
/// Uses `ObservableObject` for iOS 16 compatibility.
final class PurchaseService: ObservableObject, PurchaseProviding {

    // MARK: - Singleton

    static let shared = PurchaseService()

    // MARK: - Published Properties

    @Published private(set) var subscriptionStatus: SubscriptionStatus = .loading
    @Published private(set) var availablePlans: [SubscriptionPlan] = []
    @Published private(set) var isPurchasing: Bool = false

    // MARK: - Private Properties

    /// Cached StoreKit products keyed by product identifier.
    private var products: [String: Product] = [:]

    /// Task for observing transaction updates.
    private var transactionListenerTask: Task<Void, Never>?

    // MARK: - Initialization

    init() {
        startTransactionListener()
    }

    deinit {
        transactionListenerTask?.cancel()
    }

    // MARK: - Transaction Listener

    /// Listens for transaction updates from StoreKit (e.g., renewals, revocations).
    ///
    /// This runs for the lifetime of the service and handles:
    /// - Automatic renewals
    /// - Subscription revocations
    /// - Pending transactions completing
    private func startTransactionListener() {
        transactionListenerTask = Task.detached { [weak self] in
            for await result in Transaction.updates {
                guard let self = self else { return }
                do {
                    let transaction = try self.verifyTransaction(result)
                    await transaction.finish()
                    await self.refreshSubscriptionStatus()
                } catch {
                    #if DEBUG
                    print("[PurchaseService] Transaction update verification failed: \(error)")
                    #endif
                }
            }
        }
    }

    // MARK: - Load Products

    /// Loads available subscription products from the App Store.
    ///
    /// Fetches products for all known identifiers and converts them
    /// into `SubscriptionPlan` models with localized pricing.
    func loadProducts() async throws {
        let storeProducts: [Product]
        do {
            storeProducts = try await Product.products(for: ProductIdentifier.allSubscriptions)
        } catch {
            throw PurchaseError.storeKitError(error.localizedDescription)
        }

        guard !storeProducts.isEmpty else {
            throw PurchaseError.notAvailable
        }

        var loadedProducts: [String: Product] = [:]
        var plans: [SubscriptionPlan] = []

        for product in storeProducts {
            loadedProducts[product.id] = product

            let tier = ProductIdentifier.tier(for: product.id)
            let billingPeriod = ProductIdentifier.billingPeriod(for: product.id)
            let features = Self.features(for: tier)

            let plan = SubscriptionPlan(
                id: product.id,
                tier: tier,
                billingPeriod: billingPeriod,
                displayPrice: product.displayPrice,
                price: product.price,
                features: features,
                isRecommended: product.id == ProductIdentifier.proMonthly
            )
            plans.append(plan)
        }

        // Sort: Pro before Team, Monthly before Annual
        plans.sort { lhs, rhs in
            if lhs.tier != rhs.tier {
                return lhs.tier == .pro
            }
            return lhs.billingPeriod == .monthly
        }

        await MainActor.run {
            self.products = loadedProducts
            self.availablePlans = plans
        }
    }

    // MARK: - Purchase

    /// Purchases a subscription plan.
    ///
    /// - Parameter plan: The subscription plan to purchase.
    /// - Returns: `true` if the purchase completed successfully.
    /// - Throws: `PurchaseError` if the purchase fails.
    @discardableResult
    func purchase(_ plan: SubscriptionPlan) async throws -> Bool {
        guard let product = products[plan.id] else {
            throw PurchaseError.productNotFound(plan.id)
        }

        await MainActor.run {
            self.isPurchasing = true
        }

        defer {
            Task { @MainActor in
                self.isPurchasing = false
            }
        }

        let result: Product.PurchaseResult
        do {
            result = try await product.purchase()
        } catch let error as StoreKitError {
            throw PurchaseError.storeKitError(error.localizedDescription)
        } catch {
            throw PurchaseError.purchaseFailed(error.localizedDescription)
        }

        switch result {
        case .success(let verification):
            let transaction = try verifyTransaction(verification)
            await transaction.finish()
            await refreshSubscriptionStatus()
            return true

        case .pending:
            throw PurchaseError.pendingTransaction

        case .userCancelled:
            throw PurchaseError.purchaseCancelled

        @unknown default:
            throw PurchaseError.purchaseFailed("Unknown purchase result.")
        }
    }

    // MARK: - Restore Purchases

    /// Restores previously purchased subscriptions.
    ///
    /// Calls `AppStore.sync()` which forces StoreKit to check
    /// the user's current entitlements with the App Store server.
    func restorePurchases() async throws {
        do {
            try await AppStore.sync()
        } catch {
            throw PurchaseError.storeKitError(error.localizedDescription)
        }

        await refreshSubscriptionStatus()
    }

    // MARK: - Subscription Status

    /// Refreshes the current subscription status by checking active entitlements.
    ///
    /// Iterates through `Transaction.currentEntitlements` to find the
    /// most recent verified subscription transaction.
    func refreshSubscriptionStatus() async {
        var latestTransaction: Transaction?
        var latestProductId: String?

        for await result in Transaction.currentEntitlements {
            guard let transaction = try? verifyTransaction(result) else { continue }

            // Only consider subscription-type transactions
            guard transaction.productType == .autoRenewable else { continue }

            if let existing = latestTransaction {
                if transaction.purchaseDate > existing.purchaseDate {
                    latestTransaction = transaction
                    latestProductId = transaction.productID
                }
            } else {
                latestTransaction = transaction
                latestProductId = transaction.productID
            }
        }

        let newStatus: SubscriptionStatus

        if let transaction = latestTransaction, let productId = latestProductId {
            let tier = ProductIdentifier.tier(for: productId)
            let billingPeriod = ProductIdentifier.billingPeriod(for: productId)

            // Check subscription status details
            let statuses: [Product.SubscriptionInfo.Status]
            if let product = products[productId] {
                statuses = (try? await product.subscription?.status) ?? []
            } else {
                // Product not loaded yet, try loading it
                if let loadedProducts = try? await Product.products(for: [productId]),
                   let product = loadedProducts.first {
                    statuses = (try? await product.subscription?.status) ?? []
                } else {
                    statuses = []
                }
            }

            let willAutoRenew = statuses.first(where: { status in
                status.state == .subscribed || status.state == .inGracePeriod
            }).flatMap { status -> Bool? in
                guard case .verified(let renewalInfo) = status.renewalInfo else { return nil }
                return renewalInfo.willAutoRenew
            } ?? true

            let info = SubscriptionInfo(
                tier: tier,
                productId: productId,
                purchaseDate: transaction.purchaseDate,
                expirationDate: transaction.expirationDate,
                willAutoRenew: willAutoRenew,
                billingPeriod: billingPeriod
            )

            if let expirationDate = transaction.expirationDate, expirationDate < Date() {
                if transaction.revocationDate != nil {
                    newStatus = .revoked
                } else {
                    newStatus = .expired(info)
                }
            } else {
                newStatus = .subscribed(info)
            }
        } else {
            newStatus = .notSubscribed
        }

        await MainActor.run {
            self.subscriptionStatus = newStatus
        }
    }

    // MARK: - Transaction Verification

    /// Verifies a transaction result from StoreKit.
    ///
    /// - Parameter result: The verification result from StoreKit.
    /// - Returns: The verified transaction.
    /// - Throws: `PurchaseError.verificationFailed` if verification fails.
    private func verifyTransaction(_ result: VerificationResult<Transaction>) throws -> Transaction {
        switch result {
        case .verified(let transaction):
            return transaction
        case .unverified(_, let error):
            #if DEBUG
            print("[PurchaseService] Transaction verification failed: \(error)")
            #endif
            throw PurchaseError.verificationFailed
        }
    }

    // MARK: - Feature Definitions

    /// Returns the features for a given subscription tier.
    ///
    /// - Parameter tier: The subscription tier.
    /// - Returns: An array of `PlanFeature` describing what the tier includes.
    static func features(for tier: SubscriptionTier) -> [PlanFeature] {
        switch tier {
        case .free:
            return [
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.basicAccess", comment: ""),
                    iconName: "person.fill"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.singleDevice", comment: ""),
                    iconName: "iphone"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.limitedSessions", comment: ""),
                    iconName: "clock"
                )
            ]
        case .pro:
            return [
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.unlimitedSessions", comment: ""),
                    iconName: "infinity"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.multiDevice", comment: ""),
                    iconName: "laptopcomputer.and.iphone"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.prioritySupport", comment: ""),
                    iconName: "star.fill"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.advancedAnalytics", comment: ""),
                    iconName: "chart.bar.fill"
                )
            ]
        case .team:
            return [
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.everythingInPro", comment: ""),
                    iconName: "checkmark.seal.fill"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.teamManagement", comment: ""),
                    iconName: "person.3.fill"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.sharedSessions", comment: ""),
                    iconName: "shared.with.you"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.adminControls", comment: ""),
                    iconName: "gearshape.2.fill"
                ),
                PlanFeature(
                    title: NSLocalizedString("subscription.feature.dedicatedSupport", comment: ""),
                    iconName: "headset.circle.fill"
                )
            ]
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/enflame/happy/data/billing/PurchaseService.kt
package com.enflame.happy.data.billing

import android.app.Activity
import android.content.Context
import android.util.Log
import com.android.billingclient.api.AcknowledgePurchaseParams
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.PurchasesUpdatedListener
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.enflame.happy.domain.model.BillingPeriod
import com.enflame.happy.domain.model.PlanFeature
import com.enflame.happy.domain.model.ProductIdentifiers
import com.enflame.happy.domain.model.PurchaseResult
import com.enflame.happy.domain.model.SubscriptionInfo
import com.enflame.happy.domain.model.SubscriptionPlan
import com.enflame.happy.domain.model.SubscriptionStatus
import com.enflame.happy.domain.model.SubscriptionTier
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

/**
 * Service managing Google Play Billing for in-app subscriptions.
 *
 * Handles the full lifecycle of the billing client including connection,
 * product queries, purchase flows, acknowledgment, and subscription
 * status tracking. This is the Android counterpart to the iOS
 * `PurchaseService` (StoreKit 2).
 *
 * Key responsibilities:
 * - Connect to and disconnect from the Google Play Billing service
 * - Query available subscription products and convert to domain models
 * - Launch the purchase flow via an Activity
 * - Acknowledge purchases (required within 3 days to prevent refund)
 * - Track and refresh the current subscription status
 * - Restore previously purchased subscriptions
 *
 * All state is exposed via [StateFlow]s for reactive UI consumption.
 */
@Singleton
class PurchaseService @Inject constructor(
    @ApplicationContext private val context: Context
) {

    // MARK: - Public State

    private val _subscriptionStatus = MutableStateFlow<SubscriptionStatus>(SubscriptionStatus.Loading)

    /** The current subscription status. */
    val subscriptionStatus: StateFlow<SubscriptionStatus> = _subscriptionStatus.asStateFlow()

    private val _availablePlans = MutableStateFlow<List<SubscriptionPlan>>(emptyList())

    /** Available subscription plans loaded from Google Play. */
    val availablePlans: StateFlow<List<SubscriptionPlan>> = _availablePlans.asStateFlow()

    private val _isPurchasing = MutableStateFlow(false)

    /** Whether a purchase is currently in progress. */
    val isPurchasing: StateFlow<Boolean> = _isPurchasing.asStateFlow()

    private val _isConnected = MutableStateFlow(false)

    /** Whether the billing client is connected to Google Play. */
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    // MARK: - Private State

    /** Cached product details keyed by product ID. */
    private var productDetailsMap: Map<String, ProductDetails> = emptyMap()

    /** Callback for the current purchase flow result. */
    private var purchaseResultCallback: ((PurchaseResult) -> Unit)? = null

    // MARK: - Billing Client

    private val purchasesUpdatedListener = PurchasesUpdatedListener { billingResult, purchases ->
        handlePurchaseResult(billingResult, purchases)
    }

    private val billingClient: BillingClient = BillingClient.newBuilder(context)
        .setListener(purchasesUpdatedListener)
        .enablePendingPurchases()
        .build()

    // MARK: - Connection

    /**
     * Connects to the Google Play Billing service.
     *
     * This must be called before any billing operations. The connection
     * is maintained as a singleton and will retry on disconnection.
     *
     * @return `true` if the connection was established successfully.
     */
    suspend fun connect(): Boolean = suspendCancellableCoroutine { continuation ->
        if (billingClient.isReady) {
            _isConnected.value = true
            continuation.resume(true)
            return@suspendCancellableCoroutine
        }

        billingClient.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(billingResult: BillingResult) {
                val success = billingResult.responseCode == BillingClient.BillingResponseCode.OK
                _isConnected.value = success

                if (success) {
                    Log.d(TAG, "Billing client connected")
                } else {
                    Log.e(
                        TAG,
                        "Billing setup failed: ${billingResult.responseCode} - ${billingResult.debugMessage}"
                    )
                }

                if (continuation.isActive) {
                    continuation.resume(success)
                }
            }

            override fun onBillingServiceDisconnected() {
                _isConnected.value = false
                Log.w(TAG, "Billing service disconnected")
            }
        })
    }

    /**
     * Disconnects from the Google Play Billing service.
     *
     * Should be called when the billing service is no longer needed
     * to release resources.
     */
    fun disconnect() {
        billingClient.endConnection()
        _isConnected.value = false
        Log.d(TAG, "Billing client disconnected")
    }

    // MARK: - Query Products

    /**
     * Queries available subscription products from Google Play.
     *
     * Fetches product details for all known subscription identifiers
     * and converts them into [SubscriptionPlan] domain models with
     * localized pricing from the Play Store.
     *
     * @return `true` if products were loaded successfully.
     */
    suspend fun queryProducts(): Boolean {
        if (!ensureConnected()) return false

        val productList = ProductIdentifiers.ALL_SUBSCRIPTIONS.map { productId ->
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(productId)
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        }

        val params = QueryProductDetailsParams.newBuilder()
            .setProductList(productList)
            .build()

        val result = suspendCancellableCoroutine { continuation ->
            billingClient.queryProductDetailsAsync(params) { billingResult, productDetailsList ->
                if (continuation.isActive) {
                    continuation.resume(Pair(billingResult, productDetailsList))
                }
            }
        }

        val (billingResult, productDetailsList) = result

        if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
            Log.e(TAG, "Failed to query products: ${billingResult.debugMessage}")
            return false
        }

        if (productDetailsList.isEmpty()) {
            Log.w(TAG, "No subscription products found")
            return false
        }

        // Cache product details
        productDetailsMap = productDetailsList.associateBy { it.productId }

        // Convert to domain models
        val plans = productDetailsList.mapNotNull { details ->
            convertToSubscriptionPlan(details)
        }.sortedWith(compareBy<SubscriptionPlan> { plan ->
            // Pro before Team
            when (plan.tier) {
                SubscriptionTier.PRO -> 0
                SubscriptionTier.TEAM -> 1
                SubscriptionTier.FREE -> 2
            }
        }.thenBy { plan ->
            // Monthly before Annual
            when (plan.billingPeriod) {
                BillingPeriod.MONTHLY -> 0
                BillingPeriod.ANNUAL -> 1
            }
        })

        _availablePlans.value = plans
        Log.d(TAG, "Loaded ${plans.size} subscription plans")
        return true
    }

    // MARK: - Purchase

    /**
     * Launches the purchase flow for a subscription plan.
     *
     * This presents the Google Play purchase dialog to the user.
     * The result is returned asynchronously via the [PurchaseResult].
     *
     * @param activity The activity to launch the purchase flow from.
     * @param plan The subscription plan to purchase.
     * @return The result of the purchase attempt.
     */
    suspend fun purchase(activity: Activity, plan: SubscriptionPlan): PurchaseResult {
        if (!ensureConnected()) {
            return PurchaseResult.Failed("Billing service not connected")
        }

        val productDetails = productDetailsMap[plan.id]
            ?: return PurchaseResult.Failed("Product not found: ${plan.id}")

        // Find the matching offer (subscription plans have offers with pricing phases)
        val offerToken = plan.offerToken.ifEmpty {
            productDetails.subscriptionOfferDetails?.firstOrNull()?.offerToken
        } ?: return PurchaseResult.Failed("No offer available for product: ${plan.id}")

        val productDetailsParams = BillingFlowParams.ProductDetailsParams.newBuilder()
            .setProductDetails(productDetails)
            .setOfferToken(offerToken)
            .build()

        val billingFlowParams = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(listOf(productDetailsParams))
            .build()

        _isPurchasing.value = true

        return suspendCancellableCoroutine { continuation ->
            purchaseResultCallback = { result ->
                _isPurchasing.value = false
                if (continuation.isActive) {
                    continuation.resume(result)
                }
            }

            val billingResult = billingClient.launchBillingFlow(activity, billingFlowParams)

            if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
                _isPurchasing.value = false
                purchaseResultCallback = null
                if (continuation.isActive) {
                    continuation.resume(
                        PurchaseResult.BillingError(
                            billingResult.responseCode,
                            billingResult.debugMessage
                        )
                    )
                }
            }
        }
    }

    // MARK: - Restore Purchases

    /**
     * Restores previously purchased subscriptions.
     *
     * Queries Google Play for active subscriptions and updates
     * the subscription status accordingly.
     *
     * @return `true` if restore completed successfully.
     */
    suspend fun restorePurchases(): Boolean {
        if (!ensureConnected()) return false

        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        val result = suspendCancellableCoroutine { continuation ->
            billingClient.queryPurchasesAsync(params) { billingResult, purchases ->
                if (continuation.isActive) {
                    continuation.resume(Pair(billingResult, purchases))
                }
            }
        }

        val (billingResult, purchases) = result

        if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
            Log.e(TAG, "Failed to restore purchases: ${billingResult.debugMessage}")
            return false
        }

        processPurchases(purchases)
        return true
    }

    // MARK: - Subscription Status

    /**
     * Refreshes the current subscription status by querying Google Play.
     *
     * Checks for active subscriptions and updates the status flow.
     */
    suspend fun refreshSubscriptionStatus() {
        if (!ensureConnected()) {
            _subscriptionStatus.value = SubscriptionStatus.NotSubscribed
            return
        }

        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        val result = suspendCancellableCoroutine { continuation ->
            billingClient.queryPurchasesAsync(params) { billingResult, purchases ->
                if (continuation.isActive) {
                    continuation.resume(Pair(billingResult, purchases))
                }
            }
        }

        val (billingResult, purchases) = result

        if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
            Log.e(TAG, "Failed to query subscription status: ${billingResult.debugMessage}")
            _subscriptionStatus.value = SubscriptionStatus.Error(
                "Failed to check subscription: ${billingResult.debugMessage}"
            )
            return
        }

        updateSubscriptionStatus(purchases)
    }

    // MARK: - Private Helpers

    /**
     * Ensures the billing client is connected, attempting to connect if not.
     */
    private suspend fun ensureConnected(): Boolean {
        if (billingClient.isReady) return true
        return connect()
    }

    /**
     * Handles purchase results from the [PurchasesUpdatedListener].
     */
    private fun handlePurchaseResult(billingResult: BillingResult, purchases: List<Purchase>?) {
        val result = when (billingResult.responseCode) {
            BillingClient.BillingResponseCode.OK -> {
                if (purchases != null && purchases.isNotEmpty()) {
                    // Acknowledge purchases and update status
                    processPurchasesSync(purchases)
                    PurchaseResult.Success
                } else {
                    PurchaseResult.Failed("No purchases returned")
                }
            }

            BillingClient.BillingResponseCode.USER_CANCELED -> {
                PurchaseResult.Cancelled
            }

            BillingClient.BillingResponseCode.ITEM_ALREADY_OWNED -> {
                // User already has this subscription
                PurchaseResult.Success
            }

            BillingClient.BillingResponseCode.DEVELOPER_ERROR -> {
                Log.e(TAG, "Developer error in billing: ${billingResult.debugMessage}")
                PurchaseResult.BillingError(
                    billingResult.responseCode,
                    billingResult.debugMessage
                )
            }

            BillingClient.BillingResponseCode.NETWORK_ERROR,
            BillingClient.BillingResponseCode.SERVICE_UNAVAILABLE,
            BillingClient.BillingResponseCode.SERVICE_DISCONNECTED -> {
                _isConnected.value = false
                PurchaseResult.BillingError(
                    billingResult.responseCode,
                    billingResult.debugMessage
                )
            }

            BillingClient.BillingResponseCode.BILLING_UNAVAILABLE -> {
                PurchaseResult.BillingError(
                    billingResult.responseCode,
                    "Billing is not available on this device"
                )
            }

            BillingClient.BillingResponseCode.ITEM_UNAVAILABLE -> {
                PurchaseResult.Failed("This subscription is not available")
            }

            BillingClient.BillingResponseCode.FEATURE_NOT_SUPPORTED -> {
                PurchaseResult.Failed("Subscriptions are not supported on this device")
            }

            BillingClient.BillingResponseCode.ITEM_NOT_OWNED -> {
                PurchaseResult.Failed("Item not owned")
            }

            BillingClient.BillingResponseCode.ERROR -> {
                PurchaseResult.BillingError(
                    billingResult.responseCode,
                    billingResult.debugMessage
                )
            }

            else -> {
                PurchaseResult.BillingError(
                    billingResult.responseCode,
                    "Unknown billing error: ${billingResult.debugMessage}"
                )
            }
        }

        purchaseResultCallback?.invoke(result)
        purchaseResultCallback = null
    }

    /**
     * Processes purchases synchronously (called from listener callback).
     * Acknowledges any unacknowledged purchases.
     */
    private fun processPurchasesSync(purchases: List<Purchase>) {
        for (purchase in purchases) {
            if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED &&
                !purchase.isAcknowledged
            ) {
                acknowledgePurchase(purchase)
            }
        }
    }

    /**
     * Processes purchases from a query result (async context).
     * Acknowledges unacknowledged purchases and updates status.
     */
    private fun processPurchases(purchases: List<Purchase>) {
        for (purchase in purchases) {
            if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED &&
                !purchase.isAcknowledged
            ) {
                acknowledgePurchase(purchase)
            }
        }
        updateSubscriptionStatus(purchases)
    }

    /**
     * Acknowledges a purchase to prevent automatic refund.
     *
     * Purchases must be acknowledged within 3 days or Google Play
     * will automatically refund them.
     */
    private fun acknowledgePurchase(purchase: Purchase) {
        val params = AcknowledgePurchaseParams.newBuilder()
            .setPurchaseToken(purchase.purchaseToken)
            .build()

        billingClient.acknowledgePurchase(params) { billingResult ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                Log.d(TAG, "Purchase acknowledged: ${purchase.orderId}")
            } else {
                Log.e(
                    TAG,
                    "Failed to acknowledge purchase: ${billingResult.responseCode} - ${billingResult.debugMessage}"
                )
            }
        }
    }

    /**
     * Updates the subscription status based on current purchases.
     */
    private fun updateSubscriptionStatus(purchases: List<Purchase>) {
        // Find the most recent active subscription
        val activePurchase = purchases
            .filter { it.purchaseState == Purchase.PurchaseState.PURCHASED }
            .maxByOrNull { it.purchaseTime }

        if (activePurchase != null) {
            val productId = activePurchase.products.firstOrNull() ?: ""
            val tier = SubscriptionTier.fromProductId(productId)
            val billingPeriod = BillingPeriod.fromProductId(productId)

            val info = SubscriptionInfo(
                tier = tier,
                productId = productId,
                purchaseTimeMillis = activePurchase.purchaseTime,
                expiryTimeMillis = null, // Expiry not directly available from Purchase object
                willAutoRenew = activePurchase.isAutoRenewing,
                billingPeriod = billingPeriod
            )

            _subscriptionStatus.value = if (activePurchase.isAutoRenewing ||
                activePurchase.purchaseState == Purchase.PurchaseState.PURCHASED
            ) {
                SubscriptionStatus.Subscribed(info)
            } else {
                SubscriptionStatus.Expired(info)
            }
        } else {
            // Check for pending purchases
            val pendingPurchase = purchases.firstOrNull {
                it.purchaseState == Purchase.PurchaseState.PENDING
            }

            if (pendingPurchase != null) {
                Log.d(TAG, "Pending purchase found")
                _subscriptionStatus.value = SubscriptionStatus.NotSubscribed
            } else {
                _subscriptionStatus.value = SubscriptionStatus.NotSubscribed
            }
        }

        Log.d(TAG, "Subscription status updated: ${_subscriptionStatus.value}")
    }

    /**
     * Converts a [ProductDetails] to a [SubscriptionPlan] domain model.
     *
     * Extracts pricing from the first subscription offer's base plan
     * pricing phase.
     */
    private fun convertToSubscriptionPlan(productDetails: ProductDetails): SubscriptionPlan? {
        val subscriptionOfferDetails = productDetails.subscriptionOfferDetails ?: return null
        val offer = subscriptionOfferDetails.firstOrNull() ?: return null
        val pricingPhase = offer.pricingPhases.pricingPhaseList.firstOrNull() ?: return null

        val productId = productDetails.productId
        val tier = SubscriptionTier.fromProductId(productId)
        val billingPeriod = BillingPeriod.fromProductId(productId)

        return SubscriptionPlan(
            id = productId,
            tier = tier,
            billingPeriod = billingPeriod,
            displayPrice = pricingPhase.formattedPrice,
            priceMicros = pricingPhase.priceAmountMicros,
            currencyCode = pricingPhase.priceCurrencyCode,
            features = getFeaturesForTier(tier),
            isRecommended = productId == ProductIdentifiers.PRO_MONTHLY,
            offerToken = offer.offerToken
        )
    }

    companion object {
        private const val TAG = "PurchaseService"

        /**
         * Returns the features for a given subscription tier.
         *
         * Mirrors the iOS `PurchaseService.features(for:)` method.
         *
         * @param tier The subscription tier.
         * @return List of [PlanFeature] for the tier.
         */
        fun getFeaturesForTier(tier: SubscriptionTier): List<PlanFeature> {
            return when (tier) {
                SubscriptionTier.FREE -> listOf(
                    PlanFeature(title = "Basic access", iconName = "person"),
                    PlanFeature(title = "Single device", iconName = "phone_android"),
                    PlanFeature(title = "Limited sessions", iconName = "schedule")
                )

                SubscriptionTier.PRO -> listOf(
                    PlanFeature(title = "Unlimited sessions", iconName = "all_inclusive"),
                    PlanFeature(title = "Multiple devices", iconName = "devices"),
                    PlanFeature(title = "Priority support", iconName = "star"),
                    PlanFeature(title = "Advanced analytics", iconName = "bar_chart")
                )

                SubscriptionTier.TEAM -> listOf(
                    PlanFeature(title = "Everything in Pro", iconName = "verified"),
                    PlanFeature(title = "Team management", iconName = "groups"),
                    PlanFeature(title = "Shared sessions", iconName = "share"),
                    PlanFeature(title = "Admin controls", iconName = "admin_panel_settings"),
                    PlanFeature(title = "Dedicated support", iconName = "headset_mic")
                )
            }
        }
    }
}
